great_expectations/core/config_provider.py

```diff
@@ -1,22 +1,364 @@
 from __future__ import annotations
 
+import base64
 import errno
+import json
+import logging
 import os
+import re
 from abc import ABC, abstractmethod
 from collections import OrderedDict
-from typing import Dict, Optional, Type, cast
+from functools import lru_cache
+from typing import Any, Dict, Optional, Type, cast
 
+import great_expectations.exceptions as ge_exceptions
 from great_expectations.core.yaml_handler import YAMLHandler
-from great_expectations.data_context.types.base import GeCloudConfig
-from great_expectations.data_context.util import (
-    substitute_all_config_variables,
-    substitute_config_variable,
-)
-
+from great_expectations.data_context.types.base import BaseYamlConfig, GeCloudConfig
+
+try:
+    from azure.identity import DefaultAzureCredential
+    from azure.keyvault.secrets import SecretClient
+except ImportError:
+    SecretClient = None
+    DefaultAzureCredential = None
+
+try:
+    import boto3
+    from botocore.exceptions import ClientError
+except ImportError:
+    boto3 = None
+    ClientError = None
+
+try:
+    from google.cloud import secretmanager
+except ImportError:
+    secretmanager = None
+
+logger = logging.getLogger(__name__)
 yaml = YAMLHandler()
 
 
+class ConfigurationSubstitutor:
```

Comment:
This class is quite simply taking `substitute_all_config_variables` and `substitute_config_variable` and putting it in a class. All the logic here is ported over from `util.py`.